### PR TITLE
fix(ingest/teradata): Teradata speed up changes

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py [hunk 1]
```diff
@@ -138,11 +209,40 @@ class TeradataSource(TwoTierSQLAlchemySource):
...
+    t.LastAlterName,
+    t.LastAlterTimeStamp,
+    t.RequestText
+FROM dbc.Tables t
+WHERE DatabaseName NOT IN ('All', 'Crashdumps', 'DBC', 'dbcmngr',
+        'Default', 'External_AP', 'EXTUSER', 'LockLogShredder', 'PUBLIC',
+        'Sys_Calendar', 'SysAdmin', 'SYSBAR', 'SYSJDBC', 'SYSLIB',
+        'SystemFe', 'SYSUDTLIB', 'SYSUIF', 'TD_SERVER_DB', 'TDStats',
+        'TD_SYSGPL', 'TD_SYSXML', 'TDMaps', 'TDPUSER', 'TDQCD',
+        'tdwm', 'SQLJ', 'TD_SYSFNLIB', 'SYSSPATIAL')
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py:230]: Would be nice to reuse this list

File: metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py [hunk 2]
```diff
@@ -112,6 +161,28 @@ class TeradataConfig(BaseTeradataConfig, BaseTimeWindowConfig):
...
     )
 
+    disable_schema_metadata: bool = Field(
+        default=False,
+        description="Whether to disable schema metadata ingestion. Only table names and database names will be ingested.",
+    )
+
+    use_cached_metadata: bool = Field(
+        default=True,
+        description="Whether to use cached metadata. This reduce the number of queries to the database but requires to have cached metadata.",
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py:162]:
```suggestion
        description="Whether to use cached metadata. This reduces the number of queries to the database but requires storing all tables in memory.",
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py [hunk 3]
```diff
@@ -172,11 +272,150 @@ def __init__(self, config: TeradataConfig, ctx: PipelineContext):
...
+            self.cache_tables_and_views()
+            logger.info(f"Found {len(self._tables_cache)} tables and views")
+            setattr(self, "loop_tables", self.cached_loop_tables)  # noqa: B010
+            setattr(self, "loop_views", self.cached_loop_views)  # noqa: B010
+            setattr(  # noqa: B010
+                self, "get_table_properties", self.cached_get_table_properties
+            )
+            # self.loop_tables = self.cached_loop_tables
+            # self.loop_views = self.cached_loop_views
+            # self.get_table_properties = self.cached_get_table_properties
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py:276]:
```suggestion
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py [hunk 4]
```diff
@@ -172,11 +272,150 @@ def __init__(self, config: TeradataConfig, ctx: PipelineContext):
...
+            inspector = inspect(conn)
+            if self.config.database and self.config.database != "":
+                databases = [self.config.database]
+            else:
+                databases = inspector.get_schema_names()
+            for db in databases:
+                if self.config.database_pattern.allowed(db):
+                    # url = self.config.get_sql_alchemy_url(current_db=db)
+                    # with create_engine(url, **self.config.options).connect() as conn:
+                    #    inspector = inspect(conn)
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py:320]:
```suggestion
```

**Review by asikowitz — COMMENTED** [review]
Thanks for getting this out so fast. I'd like to hold off on merging this because it's a bit hacky -- perhaps we should refactor our SQL common connector into overridable methods instead. But great to have this

File: metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py [hunk 5]
```diff
@@ -460,15 +447,12 @@ def cache_tables_and_views(self) -> None:
...
-            self.LINEAGE_QUERY.format(
-                start_time=self.config.start_time, end_time=self.config.end_time
-            )
-        ):
+        for entry in engine.execute(self._make_lineage_query()):
             self.report.num_queries_parsed += 1
             if self.report.num_queries_parsed % 1000 == 0:
                 logger.info(f"Parsed {self.report.num_queries_parsed} queries")
-
+            print(entry.query)
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/sql/teradata.py:454]: You left this print here, I don't know if it was intentional

**Review by asikowitz — APPROVED** [review]

**Pull request merged**